Added: echoLuaErrors to profile preferences

dicene opened the pull request:
This branch adds a new checkbox under Main Display -> Display options that toggles the display of lua errors to the main console, which can help to speed up the process of tracking down and correcting errors, especially given how out-of-the-way the error screen can be for some people.

vadi2 commented:
It is deliberately out of the way - most users aren't coders, and having errors spewing up in the main window is of little practical use to them. I'm fine with this option added, with the default that it is off.

Will approve when I have some time to test it.

dicene commented:
It's set up to be off by default in that commit.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -12494,7 +12494,14 @@ void TLuaInterpreter::logError( std::string & e, const QString & name, const QSt
+    if( mpHost->mpConsole && mpHost->mEchoLuaErrors )
+    {
+    	  blue = QColor(80,80,255);
+        mpHost->mpConsole->printDebug(blue, black, QString("[ ERROR ] - ") );
```
This will look a lot like the "standardised" messages of form:
* "[ ERROR ] - Something has gone wrong"
* "[ WARN ]  - Something unusual has happened, it isn't necessarily wrong but it might be"
* "[ ALERT ] - Something unusual has happened, and this is to let the user know"
* "[ INFO ]  - Something usual has started to happen but overall the process that causes it hasn't finishing happening yet, OR something unusual has happened and this is some more information about it."
* "[  OK  ]  - Something usual has finished and things generally were all right with it."

but will not follow the same colouration.  If you __want__ to use _that_ system, just create a `QString` with a form:

    tr( "[ ERROR ] - Message beginning with a capital letter and ideally manually formatted to less than 80 characters (with some allowance for %1 type variable expansion)\n"
                    "and if split onto multiple lines with the wrapped lines indented to the first non-space character after the first lines '-' to reflect how it will show on screen\n"
                    "and with a full-stop or other end of sentence punctuation but NO line-feed on the last line (it gets stripped off!)" )
      .arg( "String inserted where-ever %1 is, possible multiple times" );

and then use the nearest available `postMessage( <above string> )` method! :grinning: 

_The colouration/formatting of the messages is handled automagically by code in:
`(void) cTelnet::postMessage( QString msg )`_

SlySven replied:
It _might_ be worth considering a new _type_ for these - possibly something like
    "[  LUA  ]  - \<error message from lua sub-system\>"
and we could make it blue if you really want it to be!

OTOH I can't be certain that all the possible sources of messages from the Lua sub-system will fit into this style AND I'm not sure if ALL the messages are error ones, it might be worth experimenting to see if various "OK" messages come from there as well...

SlySven commented on src/Host.h:
```diff
@@ -161,6 +161,7 @@ class Host  : public QObject
     bool               mAllowToSendCommand;
     bool               mAutoClearCommandLineAfterSend;
     bool               mBlockScriptCompile;
+    bool               mEchoLuaErrors; 
```
There is no initialization value set in the `Host::Host(...)` constructor, whilst the line in the `(void) XMLimport::readHostPackage( Host * pT )` code will set it to `false` when it reads the profile in, I find that `bool` variables generally do not seem to assume a `false` value (watching the class members in the debugger in the Qt Creator IDE shows random NUMBER values for them) unless explicitly set that way in the constructor.

Avoid odd effects on profile loading with a line like:
`, mEchoLuaErrors( false )`
in the list under the line containing:
`Host::Host( int port, const QString& hostname, const QString& login, const QString& pass, int id )` in the `Host.cpp` file (not-obviously) it needs to be in the same order as the member variables are declared in the `Host.h` header file as the Clang analyser will otherwise moan about the out-of-order initialisation...  {if I'm preaching to someone who already knows this then you have my apologies but just in case you weren't then now you know!} :wink:

dicene replied:
I know I had it being initialized as false at some point when I was making it, just not sure at what point I accidentally undid that change. Hrm...

SlySven replied:
Saw this has been done now in https://github.com/Mudlet/Mudlet/pull/349/commits/b32a56788c364f0683a1724caac1a20e4cbeb7a9. :+1:

keneanung commented on src/TLuaInterpreter.cpp:
```diff
@@ -12494,7 +12494,14 @@ void TLuaInterpreter::logError( std::string & e, const QString & name, const QSt
     }
-
+    
+    if( mpHost->mpConsole && mpHost->mEchoLuaErrors )
```
I'll probably want to have the error messages in the main window all the time, if I have that option set. Usually I am coding something and have the esitor window open all the time, which would remove a lot of utility of this option.

If you want to put another restriction to this option (which I wouldn't do), I'd rather do the opptosite and show them only when the editor is open, similar to the debug window.

dicene replied:
Similar to keneanung, I will have the editor open and prefer to have it echo'd to the main window regardless. I keep the editor window open 100% of the time I use Mudlet, usually on another monitor, so I think an always-on option is optimal for my purposes at least.

vadi2 replied:
I think this option should not be affected by whenever the editor is open or not - if enabled, it should always show as that's why the user enabled it to begin with.

SlySven replied:
Fair enough, disregard the comment...

dicene commented on src/mudlet.cpp:
```diff
@@ -2375,6 +2375,14 @@ void mudlet::setEditorTextoptions( const bool isTabsAndSpacesToBeShown, const bo
     emit signal_editorTextOptionsChanged( mEditorTextOptions );
 }
 
+void mudlet::setEchoLuaErrors( const bool isEchoLuaErrorsEnabled )
```
Are you saying just getting rid of the setEchoLuaErrors function and manually editing it via like pHost->mEchoLuaErrors = true/false? I thought about doing that, but was just sort of emulating the code I saw around it, since I'm coming into this with a less than thorough understanding of the interactions between the various parts of Mudlet. If it would work though, I could do that and simplify the code a bit.

SlySven replied:
Yeah, the calls _up to_ the `mudlet` class are generally for application wide settings IIRC - just set the `pHost->mEchoLuaErrors` in the place suggested.

dicene commented on src/TLuaInterpreter.cpp:
```diff
@@ -12494,7 +12494,14 @@ void TLuaInterpreter::logError( std::string & e, const QString & name, const QSt
+    
+    if( mpHost->mpConsole && mpHost->mEchoLuaErrors )
+    {
+    	  blue = QColor(80,80,255);
```
Yeah. I was changing the color slightly just because I find dark blue extremely hard to read against a black background. I'm actually thinking about adding a preference for changing the default link color at some point just to make my life a little easier. Could easily remove that if you would prefer it the standard blue color, but it looks like your suggestion below will make this irrelevant anyway.

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -769,6 +769,16 @@
+          <item row="2" column="0">
+           <widget class="QCheckBox" name="checkBox_echoLuaErrors">
+            <property name="toolTip">
+             <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Prints Lua errors to the main console.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
How about "Prints Lua errors to the main window as well as the errors view." so have a bit more clarity

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -769,6 +769,16 @@
+             <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Prints Lua errors to the main console.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
+            </property>
+            <property name="text">
+             <string>Echo Lua Errors</string>
```
Just "Echo Lua errors"

dicene replied:
Whatever changes to the text of it you want, I can absolutely whip up. Just to be sure though, you're saying the label text should stay the same as "Echo Lua errors" or should I make that more informative too(in addition to improving the tooltip).

vadi2 commented:
It doesn't sit 100% with me, but I can't think of better text. The adjusted
tooltip should clear up any confusion I hope.

On Mon, 5 Dec 2016 6:41 am dicene, <notifications@github.com> wrote:

> *@dicene* commented on this pull request.
> ------------------------------
>
> In src/ui/profile_preferences.ui
> <https://github.com/Mudlet/Mudlet/pull/349>:
>
> > @@ -769,6 +769,16 @@
>              </property>
>             </widget>
>            </item>
> +          <item row="2" column="0">
> +           <widget class="QCheckBox" name="checkBox_echoLuaErrors">
> +            <property name="toolTip">
> +             <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Prints Lua errors to the main console.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
> +            </property>
> +            <property name="text">
> +             <string>Echo Lua Errors</string>
>
> Whatever changes to the text of it you want, I can absolutely whip up.
> Just to be sure though, you're saying the label text should stay the same
> as "Echo Lua errors" or should I make that more informative too(in addition
> to improving the tooltip).
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/349>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGxjLzg9G_VSsV2UKlDpgR49AheGsLCks5rEyWPgaJpZM4Ky_SX>
> .
>

dicene commented:
Got a lot of little improvements here. Added in a way to report lua errors via the postMessage system(the one that colorizes "[  OK  ]  - " and similar messages. I'm not sure which of two ways to set that up is best.

My first inclination was to have an option for "[  LUA  ] - " style messages to be colorized like the message in the error view(blue "Error:" text on top, green "object:<> function:<>" text in the middle, and red "[script] had <> error" text on the bottom) if they were formatted to begin with something like "Lua error:", and to have normal "[  LUA ] - " style messages colorized a more normal way(like INFO) for displaying less weird coloring(just a universe blue or green or something through).

The above approach feels kind of unnecessary since any time you'd be echoing anything like that, you'd just use "[ INFO ]  - " instead, and it'd be just as useful. I can't really think of any situation where I'd want to use "[  LUA  ] - " except when echoing errors, which leads me to think I can safely assume that anything messaged as "[  LUA  ] - " is an error, and coloring it appropriately.

First option: ![http://puu.sh/sGWv2/d227d6a946.png](http://puu.sh/sGWv2/d227d6a946.png)

Second option: ![http://puu.sh/sGWyB/77233a5191.png](http://puu.sh/sGWyB/77233a5191.png)

(to be safe, the stuff that handles "[  LUA  ] - " coloring willing be coded in a manner that doesn't bug out when it encounters shorter messages than it expects, since it needs at least 3 lines to color properly)

dicene commented:
Actually, now that I've thought over it for a little while, I think the best way to handle it is to assume that anything that begins with "[  LUA  ] - " is meant to be reporting a lua error, and if it has the 3+ lines we expect(error **\n** object and function **\n** error code), color it all a generic INFO style color. That's be an efficient, properly crash-proof way to do it I think? I'll assume it's the better way unless someone specifically tells me otherwise.

dicene commented:
Also, I've adjusted the label text and tooltip of that option in the preferences window to the following:

Tooltip: Prints Lua errors to the main console in addition to the error tab in the editor.
Label: Echo Lua errors to the main console

Those seem pretty confusion free I think?

dicene commented:
Sorry to quadra post, but this feels pretty much finished to me. I've committed changes to include the copyright lines to the files I've edited, I've changed it to use a custom type of postMessage, and I've removed the setEchoLuaErrors() function from mudlet.h and mudlet.cpp both in lieu of modifying the variable directly in the host.

SlySven commented on src/XMLimport.cpp:
```diff
@@ -922,6 +923,7 @@ void XMLimport::readHostPackage( Host * pT )
     pT->mUSE_FORCE_LF_AFTER_PROMPT = ( attributes().value("mUSE_FORCE_LF_AFTER_PROMPT") == "yes" );
     pT->mUSE_UNIX_EOL = ( attributes().value("mUSE_UNIX_EOL") == "yes" );
     pT->mNoAntiAlias = ( attributes().value("mNoAntiAlias") == "yes" );
+    pT->mEchoLuaErrors = ( attributes().value("mEchoLuaErrors") == "yes" );
```
Presumably `attributes().value("mEchoLuaErrors")` returns a default constructed empty `QString` when the attribute is not present in files pre-dating the addition of this feature so that it will default to __OFF__ if the user's setup has never encountered this option before - which is what I'd expect.

SlySven commented on src/ctelnet.cpp:
```diff
@@ -1342,6 +1343,27 @@ void cTelnet::postMessage( QString msg )
+                {
+                    mpHost->mpConsole->print( body.join('\n').append('\n'), 50, 200, 50, 0, 0, 0 );  // Green on black
+                }
+            }
```
I was initially going to climb all over an addition to the `postMessage()` code but you do seem to have done a good job of plumbing it in and handling the more than two line case {body.size() > 1 at the point where you test it} and then the more than two lines case {body.size() != 0}. :+1:

SlySven commented:
I think it is good to go. :+1:

vadi2 commented:
What's the rationale behind "[  LUA  ] - Lua error: "? It's obvious it's a Lua error from the following lines anyway, and the extra line increases the screenspace it takes up.

SlySven commented:
> What's the rationale behind "[ LUA ] - Lua error: "? It's obvious it's a Lua error from the following lines anyway, and the extra line increases the screenspace it takes up.

For what it is worth the existing `(void) TLuaInterpreter::logError( std::string &, const QString &, const QString &)` command also uses __three__ lines for each error message and that is what is being reproduced in the console - and it needs the `[ something ] -`... on the first line to work the colouration code.

vadi2 commented:
Sorry, I don't see where it is using three lines for each error message - compare how terse the error log is versus the main display:

![selection_341](https://cloud.githubusercontent.com/assets/110988/21077602/ddf11e50-bf9c-11e6-8da8-09038f9a9baa.png)

SlySven commented:
BTW: currently the "run lua code" alias that is automatically included to provide an alias for `^lua (.*)$` is:

    local f,e = loadstring("return "..matches[2])
    if not f then
        f,e = assert(loadstring(matches[2]))
    end

    local r = f()
    if r ~= nil then display(r) en

given that we are moving towards making greater use of the _nil + error message_ way of handling things that fail but which are not cause by argument typing errors or other things that actually call the `lua_error( lua_State * )` to be called; I have found the following revision to the above to be useful:

    local f,e = loadstring("return "..matches[2])
    if not f then
        f,e = assert(loadstring(matches[2]))
    end

    local r,errMsg = f()
    if r ~= nil then
        display(r)
    else
        if errMsg ~= nil then
            cecho("<white:red> lua ERROR: <red:white> "..errMsg .." <reset>\n")
        end
    end

Does this have any relevance to what is being described here - if we adopt this PR will I be getting both this PR's new "[ LUA ] - Error ..." messages and something from the above revised alias (thinking about it, perhaps not - but what about `e` in the above alias, might that be the same error message?)

SlySven commented:
> Sorry, I don't see where it is using three lines for each error message ...

Um, my mistake:  yeah, I needed to look more closely at the logError(...) method to see it does use two lines. :blush:

dicene commented:
I see what you mean. While I personally like it taking up a little bit of space to increase visibility of the error message(since they like to pop up while I'm doing things that cause the screen to scroll quickly), I definitely can see the argument to reducing useless information in the message to a minimum. I think what I'll do is have the [  LUA  ] part blue, make the rest of the first line green, and cut out the "Lua error:\n" from the message, condensing it to two lines.

Only thing I'm a little iffy on is that I want to be certain that it is indicating that it's an error message, and I don't know that the data in the last line always shows that?

I'll get an updated version whipped up after I get off work tonight.

vadi2 commented:
What else could it be? Nothing else Mudlet produces looks so colourful and
has this format.

On Tue, Dec 13, 2016 at 4:27 AM dicene <notifications@github.com> wrote:

> I see what you mean. While I personally like it taking up a little bit of
> space to increase visibility of the error message(since they like to pop up
> while I'm doing things that cause the screen to scroll quickly), I
> definitely can see the argument to reducing useless information in the
> message to a minimum. I think what I'll do is have the [ LUA ] part blue,
> make the rest of the first line green, and cut out the "Lua error:\n" from
> the message, condensing it to two lines.
>
> Only thing I'm a little iffy on is that I want to be certain that it is
> indicating that it's an error message, and I don't know that the data in
> the last line always shows that?
>
> I'll get an updated version whipped up after I get off work tonight.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/349#issuecomment-266510489>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGxjMHMk06nkSS8GMg-p-F8odVXhHKgks5rHZIdgaJpZM4Ky_SX>
> .
>

dicene commented:
Well, you're right that nothing else prints anything that looks quite like it. I just brought it up because of the following two errors, one doesn't actually explicitly state that it's an error. Maybe it's just bugging me because [  LUA  ] is sort of ambiguous, but I guess it's safe to say that if someone has made a point to enable lua errors being printed to the main console, we can assume they'll know what the message is when it shows up. This look like a satisfactory finished product?

![http://puu.sh/sNAAk/8632db59ef.png](http://puu.sh/sNAAk/8632db59ef.png)

vadi2 commented:
If we'll find that we need to clarify the message, we always can. I'm happy
that that result.

On Tue, 13 Dec 2016 12:46 pm dicene, <notifications@github.com> wrote:

> Well, you're right that nothing else prints anything that looks quite like
> it. I just brought it up because of the following two errors, one doesn't
> actually explicitly state that it's an error. Maybe it's just bugging me
> because [ LUA ] is sort of ambiguous, but I guess it's safe to say that if
> someone has made a point to enable lua errors being printed to the main
> console, we can assume they'll know what the message is when it shows up.
> This look like a satisfactory finished product?
>
> [image: http://puu.sh/sNAAk/8632db59ef.png]
> <https://camo.githubusercontent.com/02ade574a433b6508eb6672f5e4467146058d09e/687474703a2f2f7075752e73682f734e41416b2f383633326462353965662e706e67>
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/349#issuecomment-266620339>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGxjD2eYouN0d6c7NW2BKUgcoUXSLNDks5rHgcJgaJpZM4Ky_SX>
> .
>

dicene commented:
Awesome. I'll run another test to make sure it works with any number of lines just in case, and I'll clean up the code a little(removing some comments that aren't necessary).

vadi2 commented:
Thanks @dicene for the improvement and @SlySven, @keneanung for contributing to the review!

vadi2 commented:
Ah man, I'm a shitty tester:

![selection_001](https://cloud.githubusercontent.com/assets/110988/21219344/2430fa14-c300-11e6-9202-fa08c70b2c58.png)

Think we should have it insert a new line if there's any text on the line the cursor is in.

dicene commented:
Haha. I'll see what I can whip up. Any idea if there is an example I can
steal from in Mudlet already?

On Dec 15, 2016 04:53, "Vadim Peretokin" <notifications@github.com> wrote:

> Ah man, I'm a shitty tester:
>
> [image: selection_001]
> <https://cloud.githubusercontent.com/assets/110988/21219344/2430fa14-c300-11e6-9202-fa08c70b2c58.png>
>
> Think we should have it insert a new line if there's any text on the line
> the cursor is in.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/349#issuecomment-267286371>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AOXs6noJGBUNv67PBKf5xCvUYNnD6BnRks5rIQ4kgaJpZM4Ky_SX>
> .
>

vadi2 commented:
Don't think so

On Thu, 15 Dec 2016 11:38 pm dicene, <notifications@github.com> wrote:

> Haha. I'll see what I can whip up. Any idea if there is an example I can
> steal from in Mudlet already?
>
> On Dec 15, 2016 04:53, "Vadim Peretokin" <notifications@github.com> wrote:
>
> > Ah man, I'm a shitty tester:
> >
> > [image: selection_001]
> > <
> https://cloud.githubusercontent.com/assets/110988/21219344/2430fa14-c300-11e6-9202-fa08c70b2c58.png
> >
> >
> > Think we should have it insert a new line if there's any text on the line
> > the cursor is in.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Mudlet/Mudlet/pull/349#issuecomment-267286371>, or
> mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AOXs6noJGBUNv67PBKf5xCvUYNnD6BnRks5rIQ4kgaJpZM4Ky_SX
> >
> > .
> >
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/349#issuecomment-267329018>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGxjEiWAOSZQ_K0YCMjDEm7aw7HeMJrks5rIULBgaJpZM4Ky_SX>
> .
>

SlySven commented:
With regard to new lines - there are always going to be issues with user/script output as some scripters like to only __begin__ any messages they put onto a console with a linefeed - and some to only __end__ their same messages.  The former is more useful if you are building up an output line a bit at a time but the latter is more conventional.  Inserting error messages into the mix is not going to look pretty for everyone however it is done - IMHO.

Given that this PR had been merged, remember @dicene that you will need to start a new branch on your repository and use that to create a new PR to _fix_ anything - you cannot _I believe_ post new commits onto this branch. 😞

vadi2 commented:
Why wouldn't you be able to check if the line the cursor is on has any text
in it already, and if it does, add a line break?

On Fri, 16 Dec 2016 6:30 am Stephen Lyons, <notifications@github.com> wrote:

> With regard to new lines - there are always going to be issues with
> user/script output as some scripters like to only *begin* any messages
> they put onto a console with a linefeed - and some to only *end* their
> same messages. The former is more useful if you are building up an output
> line a bit at a time but the latter is more conventional. Inserting error
> messages into the mix is not going to look pretty for everyone however it
> is done - IMHO.
>
> Given that this PR had been merged, remember @dicene
> <https://github.com/dicene> that you will need to start a new branch on
> your repository and use that to create a new PR to *fix* anything - you
> cannot *I believe* post new commits onto this branch. 😞
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/349#issuecomment-267435768>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGxjGsx4Ol14SpaB1Bpcix3p-7MNM13ks5rIaN-gaJpZM4Ky_SX>
> .
>

SlySven commented:
I'm not that hot on accessing the TConsole contents and I was not sure how simple it would be to "..._check if the line the cursor is on has any text in it already_..." it may be trivially easy I just don't know myself!

dicene commented:
Pretty sure I'll be able to figure out a way to read the buffer's last line from the LuaInterpreter(since that's where the error message is coming from), just gonna take a bit of trial and error. Then it shouldn't be hard to check and see if the last character in it is a new-line, and if not, add my own. I'll try and get it figured out when I can.
